posts/2020-03-10-calculating-pi-concurrently.adoc

```diff
@@ -0,0 +1,926 @@
+        Helpers.printMsg(s"|Delta|: ${delta(new MathContext(8)).abs}")
+        Helpers.printCalculationTime(startTime, stopTime)
+        fjPool.shutdown()
+      case Failure(e) =>
+        println(s"An error occurred: ${e}")
+        fjPool.shutdown()
+    }
+
+----
+
+Note that the number of threads in the `ForkJoinPool` is obtained from a setting in
+object `Settings`. The actual value can be set as a configuration setting
+(`calculating-pi.parallelism`)
+
+When this program is executed on my laptop (a MacBook Pro with a 2,6 GHz 6-Core
+Intel Core i7 processor), it produces the following output (values of π truncated):
+
+----
+man [e] > calculating-pi-concurrently > calculating pi with futures > run 3000 4000
+[info] running com.lunatech.pi.Main 3000 4000
+Iteration count = 3000 - Precision = 4000
+Calculating π with 3008 terms in 64 chunks of 47 terms each
+Threadpool size: 12
+BigDecimal precision settings: precision=4000 roundingMode=HALF_UP
+Started @ offset: 0
+Started @ offset: 94
+Started @ offset: 141
+Started @ offset: 47
+Started @ offset: 188
+Started @ offset: 235
+Started @ offset: 282
+Started @ offset: 329
+Started @ offset: 423
+Started @ offset: 376
+Started @ offset: 470
+Started @ offset: 517
+[success] Total time: 1 s, completed 19 Apr 2021, 18:51:35
+Started @ offset: 564
+Started @ offset: 611
+Started @ offset: 658
+Started @ offset: 705
+Started @ offset: 752
+Started @ offset: 799
+Started @ offset: 846
+Started @ offset: 893
+Started @ offset: 940
+Started @ offset: 987
+Started @ offset: 1034
+Started @ offset: 1081
+Started @ offset: 1128
+Started @ offset: 1175
+Started @ offset: 1222
+Started @ offset: 1269
+Started @ offset: 1316
+Started @ offset: 1363
+Started @ offset: 1410
+Started @ offset: 1457
+Started @ offset: 1504
+Started @ offset: 1551
+Started @ offset: 1598
+Started @ offset: 1645
+Started @ offset: 1692
+Started @ offset: 1739
+Started @ offset: 1786
+Started @ offset: 1833
+Started @ offset: 1880
+Started @ offset: 1927
+Started @ offset: 1974
+Started @ offset: 2021
+Started @ offset: 2068
+Started @ offset: 2115
+Started @ offset: 2162
+Started @ offset: 2209
+Started @ offset: 2256
+Started @ offset: 2303
+Started @ offset: 2350
+Started @ offset: 2397
+Started @ offset: 2444
+Started @ offset: 2491
+Started @ offset: 2538
+Started @ offset: 2585
+Started @ offset: 2632
+Started @ offset: 2679
+Started @ offset: 2726
+Started @ offset: 2773
+Started @ offset: 2820
+Started @ offset: 2867
+Started @ offset: 2914
+Started @ offset: 2961
+Pi:      3.141592653589793238462643383279502884197169...
+|Delta|: 2.8076968E-3630
+Calculation time: 1.345
+
+----
+
+What we can observe is that, with 3,008 terms, we have correctly calculated more
+than 3,600 digits accurately.
+
+If we set the size of the ForkJoinPool to 1, 2, 4, 8, 12, 16, 32 we get the
+following values for the calculation time (in seconds) respectively: 12.12,
+6.23, 3.03, 1.78, 1.34, 1.36, 1.43
```

Comment:
Timings like these would be much nicer in a graph or at least a table.

Reply:
Good point! I'll make a small graph of these numbers